### DISCUSSION TITLE
updates of versions specified in `npx` commands in CI workflow definitions

### DISCUSSION THREAD

**Discussion by travi** [renovatebot/renovate#12029] [Suggest an Idea]
i don't believe this is currently supported, but please point me in the right direction if i am incorrect. 

### Idea

has it been considered to enable updating a package version that is specified in an `npx` command within a CI workflow definition (like `npx semantic-release@18`)? this may apply to more situations, but i'm currently thinking about this for github actions workflow files. similar to how the [github-actions manager](https://docs.renovatebot.com/modules/manager/github-actions/) is able to update action versions, i wonder if the `npm` manager could also update versions specified in `npx` commands in those same workflow files.

### Context

in an effort to better communicate our recommended usage of `semantic-release`, and based on [some feedback](https://github.com/semantic-release/semantic-release/issues/2140) about our recent major release, we are thinking through the trade-offs we want to communicate between various installation methods. 

from a simplified context, the primary options that we plan to communicate are:
* early (local) installation specified through a dev-dependency in the `package.json`
* late (whether global or local) installation done just before or as part of the execution of the release process

the major trade-offs of these options are:
early:
* version is specified in the `package.json` and can be managed with tools like renovate (+)
* `semantic-release` is a release tool intended to be used in a CI pipeline, so defining the dependency in the `package.json` would result in `semantic-release` and its dependencies being installed anywhere that the project's dependencies are installed, like every engineer's machine, even though they are doing no release tasks (-)
* we define `npm` as a dependency, which is sizable. also, having `npm` in the dependency tree can conflict with local usage of the `npm` cli (-)

late:
* `semantic-release` and its dependencies are only installed where the release happens (in the CI pipeline). this reduces download size for engineers and avoids conflicts with use of the `npm` cli (+)
* the installed version when using `npm` commands or `npx` in the CI workflow defaults to installing the latest version. it is debatable how risky this is but does leave open the potential of a surprise (-)
* it is possible to specify a version range when installing, which mitigates the risk of being surprised by the timing of breaking changes, but then specifies the version in a location that is even less likely to be updated than when specified in the `package.json`, raising the risk of using an older version. (-)

### A Potentially Better Way

we plan to recommend using the latest LTS version of node (we've already updated this part of our recommendation, thanks @HonkingGoose for the contribution) for the release phase and installing late using `npx semantic-release` with no version specified. we will make the trade-offs clear, but think this is the best option in most cases.

if renovate was able to update `npx semantic-release@18` to `npx semantic-release@19` in a GitHub Actions workflow when the next major version was released, it could give us an even safer alternative to offer as an option in our docs. `npx` would be a big win, but i imagine some may also want this to work with `npm`/`yarn` install commands.

would it be reasonable to support behavior like this? has this already been considered?

**Comment by viceice** [comment]
This should be easily doable with a regex manager.

**Reply by rarkins** [reply]
Yes, a regex manager script to detect `npx [package name]@currentValue` and point it to npm datasource should be enough.

**Reply by travi** [reply]
thanks to you both. will give it a shot and report when we have a chance to confirm.

**Reply by HonkingGoose** [reply]
@travi You might take a look at this for inspiration: [regex managers from `renovate/.github/default.json`](https://github.com/renovatebot/.github/blob/ae87eb94326f99f859f6e0b8707c69f5642e691a/default.json#L98-L139)

**Reply by travi** [reply]
long overdue, but i did finally circle back to confirm the approach and got it added to our docs: https://github.com/semantic-release/semantic-release/pull/2786. 

thanks for the info!
